internal/handlers/tigris/msg_aggregate.go

```diff
@@ -19,36 +19,134 @@ import (
 	"github.com/FerretDB/FerretDB/internal/util/must"
 	"github.com/FerretDB/FerretDB/internal/wire"
 )
 
 // MsgAggregate implements HandlerInterface.
 func (h *Handler) MsgAggregate(ctx context.Context, msg *wire.OpMsg) (*wire.OpMsg, error) {
-	document, err := msg.Document()
+	dbPool, err := h.DBPool(ctx)
 	if err != nil {
 		return nil, lazyerrors.Error(err)
 	}
 
-	pipeline, err := common.GetRequiredParam[*types.Array](document, "pipeline")
+	document, err := msg.Document()
 	if err != nil {
+		return nil, lazyerrors.Error(err)
+	}
+
+	// TODO https://github.com/FerretDB/FerretDB/issues/1892
+	common.Ignored(document, h.L, "cursor", "lsid")
+
+	if err = common.Unimplemented(document, "explain", "bypassDocumentValidation", "hint"); err != nil {
+		return nil, err
+	}
+
+	if err = common.Unimplemented(document, "readConcern", "writeConcern"); err != nil {
+		return nil, err
+	}
+
+	if err = common.Unimplemented(document, "let"); err != nil {
 		return nil, err
 	}
 
-	// TODO https://github.com/FerretDB/FerretDB/issues/1891
+	common.Ignored(document, h.L, "allowDiskUse", "maxTimeMS", "collation", "comment")
+
+	var qp tigrisdb.QueryParams
+
+	qp.DB, err = common.GetRequiredParam[string](document, "$db")
+	if err != nil {
+		return nil, err
+	}
 
-	if pipeline.Len() > 0 {
-		d := must.NotFail(pipeline.Get(0)).(*types.Document)
+	collection, err := document.Get(document.Command())
+	if err != nil {
+		return nil, err
+	}
 
+	// TODO handle collection-agnostic pipelines ({aggregate: 1})
+	// https://github.com/FerretDB/FerretDB/issues/1890
+	var ok bool
+	if qp.Collection, ok = collection.(string); !ok {
 		return nil, commonerrors.NewCommandErrorMsgWithArgument(
-			commonerrors.ErrNotImplemented,
-			fmt.Sprintf("`aggregate` %q is not implemented yet", d.Command()),
-			d.Command(),
+			commonerrors.ErrBadValue,
+			fmt.Sprintf("collection name has invalid type %s", common.AliasFromType(collection)),
+			document.Command(),
 		)
 	}
 
-	return nil, commonerrors.NewCommandErrorMsg(commonerrors.ErrNotImplemented, "`aggregate` command is not implemented yet")
+	pipeline, err := common.GetRequiredParam[*types.Array](document, "pipeline")
+	if err != nil {
+		return nil, err
+	}
+
+	stagesDocs := must.NotFail(iterator.Values(pipeline.Iterator()))
+	stages := make([]aggregations.Stage, len(stagesDocs))
+
+	for i, d := range stagesDocs {
+		d, ok := d.(*types.Document)
+		if !ok {
+			return nil, commonerrors.NewCommandErrorMsgWithArgument(
+				commonerrors.ErrBadValue,
+				"Each element of the 'pipeline' array must be an object",
+				document.Command(),
+			)
+		}
+
+		var s aggregations.Stage
+
+		if s, err = aggregations.NewStage(d); err != nil {
+			return nil, err
+		}
+
+		stages[i] = s
+	}
+
+	// TODO pushdown `$match` https://github.com/FerretDB/FerretDB/issues/1894
+
+	var docs []*types.Document
+
+	iter, err := dbPool.QueryDocuments(ctx, &qp)
+	if err != nil {
+		return nil, err
+	}
+
+	defer iter.Close()
```

Comment:
Are there any potential issues that would result from calling `Close` function twice? `iterator.Values()` will call `Close` at the end too.
If we left `Close` here just to be sure, and it doesn't affect anything then it's ok I guess!

Reply:
`Close` call is idempotent, see the documentation :hugs: 

> Close must be concurrency-safe and may be called multiple times.
All calls after the first should have no observable effect.

Reply:
If it is not idempotent, that's a bug we should fix. Future changes for cursors and aggregation pipelines will require Close to be safe to be called anytime.